src/snowflake/connector/nanoarrow_cpp/ArrowIterator/CArrowTableIterator.cpp

```diff
@@ -0,0 +1,844 @@
+
+        case SnowflakeType::Type::TIMESTAMP_NTZ:
+        {
+          int scale = 9;
+          if(metadata != nullptr) {
+            struct ArrowStringView scaleString = ArrowCharView(nullptr);
+            returnCode = ArrowMetadataGetValue(metadata, ArrowCharView("scale"), &scaleString);
+            SF_CHECK_ARROW_RC(returnCode, "[Snowflake Exception] error getting 'scale' from Arrow metadata, error code: %d", returnCode);
+            scale = std::stoi(std::string(scaleString.data, scaleString.size_bytes));
+          }
+          convertTimestampColumn_nanoarrow(batchIdx, colIdx, &columnSchemaView, columnArray, scale);
+          break;
+        }
+
+        case SnowflakeType::Type::TIMESTAMP_LTZ:
+        {
+            int scale = 9;
+            if(metadata != nullptr) {
+                struct ArrowStringView scaleString = ArrowCharView(nullptr);
+                returnCode = ArrowMetadataGetValue(metadata, ArrowCharView("scale"), &scaleString);
+                SF_CHECK_ARROW_RC(returnCode, "[Snowflake Exception] error getting 'scale' from Arrow metadata, error code: %d", returnCode);
+                scale = std::stoi(std::string(scaleString.data, scaleString.size_bytes));
+            }
+
+          convertTimestampColumn_nanoarrow(batchIdx, colIdx, &columnSchemaView, columnArray, scale,  m_timezone);
+          break;
+        }
+
+        case SnowflakeType::Type::TIMESTAMP_TZ:
+        {
+            struct ArrowStringView scaleString = ArrowCharView(nullptr);
+            struct ArrowStringView byteLengthString = ArrowCharView(nullptr);
+            int scale = 9;
+            int byteLength = 16;
+            if(metadata != nullptr) {
+                returnCode = ArrowMetadataGetValue(metadata, ArrowCharView("scale"), &scaleString);
+                SF_CHECK_ARROW_RC(returnCode, "[Snowflake Exception] error getting 'scale' from Arrow metadata, error code: %d", returnCode);
+                returnCode = ArrowMetadataGetValue(metadata, ArrowCharView("byteLength"), &byteLengthString);
+                SF_CHECK_ARROW_RC(returnCode, "[Snowflake Exception] error getting 'byteLength' from Arrow metadata, error code: %d", returnCode);
+                scale = std::stoi(std::string(scaleString.data, scaleString.size_bytes));
+                byteLength = std::stoi(std::string(byteLengthString.data, byteLengthString.size_bytes));
+            }
+
+          convertTimestampTZColumn_nanoarrow(batchIdx, colIdx, &columnSchemaView, columnArray, scale, byteLength, m_timezone);
+          break;
+        }
+
+        default:
+        {
+            std::string errorInfo = Logger::formatString(
+                "[Snowflake Exception] unknown snowflake data type : %s",
+                snowflakeLogicalType.data);
+            logger->error(__FILE__, __func__, __LINE__, errorInfo.c_str());
+            PyErr_SetString(PyExc_Exception, errorInfo.c_str());
+            return;
+        }
+      }
+    }
+    m_tableConverted = true;
```

Comment:
Do we not need `needsRebuild` any more?

Reply:
yes, the implementation is different here -- the previous implementation did in-place replacement while the nanaorrow version create new array schema.